src/Features/LanguageServer/Protocol/Handler/Configuration/DidChangeConfigurationNotificationHandler.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Configuration
+{
+    internal partial class DidChangeConfigurationNotificationHandler : ILspServiceNotificationHandler<LSP.DidChangeConfigurationParams>, IOnInitialized
+    {
+        private readonly ILspLogger _lspLogger;
+        private readonly IGlobalOptionService _globalOptionService;
+        private readonly IClientLanguageServerManager _clientLanguageServerManager;
+        private readonly Guid _registrationId;
+
+        public DidChangeConfigurationNotificationHandler(ILspLogger logger, IGlobalOptionService globalOptionService, IClientLanguageServerManager clientLanguageServerManager)
+        {
+            _lspLogger = logger;
+            _globalOptionService = globalOptionService;
+            _clientLanguageServerManager = clientLanguageServerManager;
+            _registrationId = Guid.NewGuid();
+        }
+
+        public bool MutatesSolutionState => false;
+
+        public bool RequiresLSPSolution => true;
+
+        [LanguageServerEndpoint(Methods.WorkspaceDidChangeConfigurationName)]
+        public Task HandleNotificationAsync(DidChangeConfigurationParams request, RequestContext requestContext, CancellationToken cancellationToken)
+            => RefreshOptionsAsync(cancellationToken);
+
+        private async Task RefreshOptionsAsync(CancellationToken cancellationToken)
+        {
+            var globalConfigurationItems = s_supportedGlobalOptions.SelectAsArray(option => new ConfigurationItem() { ScopeUri = null, Section = option.Definition.ConfigName });
+            var perLanguageConfigurationItems = s_supportedPerLanguageOptions.SelectAsArray(option => new ConfigurationItem() { ScopeUri = null, Section = option.Definition.ConfigName });
+            var allConfigurationItems = globalConfigurationItems.Concat(perLanguageConfigurationItems);
+
+            var configurationsFromClient = await GetConfigurationsAsync(allConfigurationItems, cancellationToken).ConfigureAwait(false);
```

Comment:
The DidChangeConfigurationParams should already tell us which settings are updated - so we shouldn't need to go ask the client again for it.

Basically the way I'd expect this to work is
1.  In OnInitializedAsync - we make a request from the server for the initial set of settings for the whole workspace using `WorkspaceConfigurationName`
2.  After we've asked for them once, when the settings change - the client sends us a `WorkspaceDidChangeConfigurationName` with the settings change.  We just update our internal store here.

And in a future PR we can consider how per document settings work.

Reply:
answered offline - this is actually the recommended pattern.  Ignore my previous comment.